drivers/modem/ublox-sara-r4.c

```diff
@@ -0,0 +1,1854 @@
+	} else {
+		k_sem_reset(&sock->sock_send_sem);
+		ret = k_sem_take(&sock->sock_send_sem, timeout);
+	}
+
+	if (ret == 0) {
+		ret = ictx.last_error;
+	} else if (ret == -EAGAIN) {
+		ret = -ETIMEDOUT;
+	}
+
+	return ret;
+}
+
+static int send_data(struct modem_socket *sock,
+		     const struct sockaddr *dst_addr, int dst_port,
+		     struct net_pkt *pkt)
+{
+	int ret, i;
+	struct net_buf *frag;
+	char buf[sizeof("AT+USO**=#,!###.###.###.###!,#####,####\r\n")];
+
+	if (!sock) {
+		return -EINVAL;
+	}
+
+	ictx.last_error = 0;
+
+	frag = pkt->frags;
+
+	/* use SOCKWRITE with binary mode formatting */
+	if (sock->ip_proto == IPPROTO_UDP) {
+		snprintk(buf, sizeof(buf), "AT+USOST=%d,\"%s\",%u,%u\r\n",
+			 sock->socket_id, modem_sprint_ip_addr(dst_addr),
+			 dst_port, net_buf_frags_len(frag));
+	} else {
+		snprintk(buf, sizeof(buf), "AT+USOWR=%d,%u\r\n",
+			 sock->socket_id, net_buf_frags_len(frag));
+	}
+	mdm_receiver_send(&ictx.mdm_ctx, buf, strlen(buf));
+
+	/* Slight pause per spec so that @ prompt is received */
+	k_sleep(MDM_PROMPT_CMD_DELAY);
+
+	/* Loop through packet data and send */
+	while (frag) {
+		if (frag->len > 0) {
+			/*
+			 * HACK: Apparently enabling HEX transmit mode also
+			 * affects the BINARY send method.  We need to encode
+			 * the "binary" data as HEX values here
+			 * TODO: Something faster here.
+			 */
+			for (i = 0; i < frag->len; i++) {
+				snprintk(buf, sizeof(buf), "%02x", frag->data[i]);
+				mdm_receiver_send(&ictx.mdm_ctx, buf, 2);
+			}
+		}
+		frag = frag->frags;
+	}
+
+	k_sem_reset(&sock->sock_send_sem);
+	ret = k_sem_take(&sock->sock_send_sem, MDM_CMD_SEND_TIMEOUT);
+	if (ret == 0) {
+		ret = ictx.last_error;
+	} else if (ret == -EAGAIN) {
+		ret = -ETIMEDOUT;
+	}
+
+	return ret;
+}
+
+/*** NET_BUF HELPERS ***/
+
+static bool is_crlf(u8_t c)
+{
+	if (c == '\n' || c == '\r') {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+static void net_buf_skipcrlf(struct net_buf **buf)
+{
+	/* chop off any /n or /r */
+	while (*buf && is_crlf(*(*buf)->data)) {
+		net_buf_pull_u8(*buf);
+		if (!(*buf)->len) {
+			*buf = net_buf_frag_del(NULL, *buf);
+		}
+	}
+}
+
+static u16_t net_buf_findcrlf(struct net_buf *buf, struct net_buf **frag,
+			      u16_t *offset)
+{
+	u16_t len = 0U, pos = 0U;
+
+	while (buf && !is_crlf(*(buf->data + pos))) {
+		if (pos + 1 >= buf->len) {
+			len += buf->len;
+			buf = buf->frags;
+			pos = 0U;
+		} else {
+			pos++;
+		}
+	}
+
+	if (buf && is_crlf(*(buf->data + pos))) {
+		len += pos;
+		*offset = pos;
+		*frag = buf;
+		return len;
+	}
+
+	return 0;
+}
+
+/*** UDP / TCP Helper Function ***/
+
+/* Setup IP header data to be used by some network applications.
+ * While much is dummy data, some fields such as dst, port and family are
+ * important.
+ * Return the IP + protocol header length.
+ */
+static int pkt_setup_ip_data(struct net_pkt *pkt, struct modem_socket *sock,
+			     struct sockaddr *src)
+{
+	int hdr_len = 0;
+	u16_t src_port = 0U;
```

Comment:
Keeping src_port as 0 causes DTLS packets to be dropped by `dtls_is_peer_addr_valid`. I managed to get a DTLS connection established and exchanging appdata with `src_port = net_sin(&sock->dst)->sin_port;` in the AF_INET case near line 486. Note the reversal of src/dst since these packets come from dst. I imagine that the AF_INET6 case is similar, but I currently can't test that one.

Reply:
@weinholtendian: I'll take a look at this.. I didn't test DTLS yet on the modem.